delfin/drivers/ibm/storwize_svc/ssh_handler.py

```diff
@@ -0,0 +1,363 @@
+        try:
+            if command_str is not None and ssh is not None:
+                stdin, stdout, stderr = ssh.exec_command(command_str)
+                res, err = stdout.read(), stderr.read()
+                re = res if res else err
+                result = re.decode()
+        except paramiko.AuthenticationException as ae:
+            LOG.error('doexec Authentication error:{}'.format(ae))
+            raise exception.InvalidUsernameOrPassword()
+        except Exception as e:
+            LOG.error('doexec InvalidUsernameOrPassword error:{}'.format(e))
+            if 'timed out' in str(e):
+                raise exception.SSHConnectTimeout()
+            elif 'No authentication methods available' in str(e) \
+                    or 'Authentication failed' in str(e):
+                raise exception.SSHInvalidUsernameOrPassword()
+            elif 'not a valid RSA private key file' in str(e):
+                raise exception.InvalidPrivateKey()
+            elif 'not found in known_hosts' in str(e):
+                raise exception.SSHNotFoundKnownHosts(self.ssh_pool.ssh_host)
+            else:
+                raise exception.SSHException()
+        return result
+
+    def handle_detail_ssh(self, context, command, method):
+        try:
+            with self.ssh_pool.item() as ssh:
+                ssh_info = self.do_exec(command, ssh)
+            return ssh_info
+        except Exception as e:
+            msg = "Failed to ssh ibm storwize_svc %s: %s" % \
+                  (method, six.text_type(e))
+            raise exception.SSHException(msg)
+
+    def handle_capacity(self, value):
+        capacity = 0
+        if value:
+            if value.isdigit():
+                capacity = float(value)
+            else:
+                unit = value[-2:]
+                capacity = float(value[:-2]) * int(self.units_to_bytes(unit))
+        return capacity
+
+    def get_storage(self, context):
+        try:
+            syste_minfo = self.handle_detail_ssh(context,
+                                                 self.systeminfo_command,
+                                                 'storage')
+            enclosure_info = self.handle_detail_ssh(context,
+                                                    self.enclosure_command,
+                                                    'enclosure')
+            enclosure_res = enclosure_info.split('\n')
+            enclosure = enclosure_res[1].split(':')
+            serial_number = enclosure[7]
+            map = {}
+            self.handle_detail(syste_minfo, map, split=' ')
+
+            status = map.get('statistics_status') == 'on' and 'normal' or \
+                'offline'
+            location = map.get('location')
+            free_capacity = self.handle_capacity(map.get(
+                'total_free_space'))
+            used_capacity = self.handle_capacity(map.get(
+                'total_used_capacity'))
+            raw_capacity = self.handle_capacity(map.get(
```

Comment:
SVC has internal capacity and external capacity.. Are we taking only internal capacity as raw_capacity?
I think we should clearly define how internal and external capacities are calculated.. otherwize it will be double counting at clients end if both external array and SVC are discovered.

Reply:
there is no internal capacity and external capacity,just have total capacity

Reply:
Probabaly the device you are using is not having external capacity , but the SVC architecture supports external cpacity. SVC is basically to virtualize storage from mutiple physical arrays..